components/script/dom/bluetoothpermissionresult.rs

```diff
@@ -0,0 +1,219 @@
+        self.global().as_window().bluetooth_thread()
+    }
+
+    // https://w3c.github.io/permissions/#create-a-permissionstatus
+    pub fn create_from_descriptor(global: &GlobalScope,
+                                  descriptor: PermissionDescriptorType)
+                                  -> Root<BluetoothPermissionResult> {
+        let bt_permission_result = BluetoothPermissionResult::new(global);
+        bt_permission_result.upcast::<PermissionStatus>().set_query(descriptor);
+        bt_permission_result
+    }
+
+    #[allow(unrooted_must_root)]
+    // https://webbluetoothcg.github.io/web-bluetooth/#query-the-bluetooth-permission
+    pub fn permission_query(&self, promise: &Rc<Promise>, cx: *mut JSContext) {
+        let descriptor = self.parent.get_query().borrow();
+        let bluetoth_descriptor = match *descriptor {
+            PermissionDescriptorType::Bluetooth(ref d) => d,
+            _ => return promise.reject_error(cx, (Error::Type(DESCRIPTOR_TYPE_ERROR.to_owned()))),
+        };
+
+        // Step 2.
+        self.parent.permission_query(&*descriptor);
+
+        // Step 3.
+        if let PermissionState::Denied = self.parent.State() {
+            *self.devices.borrow_mut() = Vec::new();
+            return promise.resolve_native(cx, &self.parent);
+        }
+
+        // Step 4.
+        let mut matching_devices: Vec<JS<BluetoothDevice>>  = Vec::new();
+
+        // TODO: Step 5: Create a map between the current setting object and BluetoothPermissionData
+        // extra permission data, which replaces the exisitng EXTRA_PERMISSION_DATA global variable.
+        // For this also use the extra permission data constraints from the specification:
+        // https://webbluetoothcg.github.io/web-bluetooth/#dictdef-bluetoothpermissiondata
+
+        let bluetooth = self.get_bluetooth();
+        let device_map = bluetooth.get_device_map().borrow();
+        let mut device_id = String::new();
+
+        // Step 5.
+        let allowed_devices = get_allowed_devices();
+
+        // Step 6.
+        for allowed_device in allowed_devices {
+            // Step 6.1.
+            if let Some(id) = bluetoth_descriptor.deviceId.clone() {
+                if allowed_device.deviceId != id {
+                    continue;
+                } else {
+                    device_id = String::from(id);
+                }
+            }
+
+            // Step 6.2.
+            // Instead of creating an internal slot we send an ipc message to the Bluetooth thread
+            // to check if one of the filters matches.
+            if let Some(ref filters) = bluetoth_descriptor.filters {
+                let mut scan_filters: Vec<BluetoothScanfilter> = Vec::new();
+                // TODO: Create an issue for the spec, to make the canonicalization step here.
+                for filter in filters {
+                    match canonicalize_filter(&filter) {
+                        Ok(f) => scan_filters.push(f),
```

Comment:
Applying an iterator to `filters` which also rejects the promise only with the first error, and returns
when there was an error results the following code:
```rust
 let scan_filters = filters.into_iter().map(|ref f| canonicalize_filter(f).unwrap_or_else(|err| {
    if !promise.is_settled() {
        promise.reject_error(cx, err);
    }
    BluetoothScanfilter::new_empty()
})).collect();
let scan_filter_sequence = BluetoothScanfilterSequence::new(scan_filters);
if scan_filter_sequence.has_empty_or_invalid_filter() {
    return;
}
```

This is way to longer and complicated than the current solution:
```rust
let mut scan_filters: Vec<BluetoothScanfilter> = Vec::new();
for filter in filters {
    match canonicalize_filter(&filter) {
        Ok(f) => scan_filters.push(f),
        Err(err) => {
            return promise.reject_error(cx, err);
        },
    }
}
```